Clean up IP header dependencies in TCP2

pfl opened the pull request:
Use existing functions to compute the start of the TCP header. Add a check on CONFIG_HEAP_MEM_POOL_SIZE in order to have memory for k_{c,m}alloc used by tcp2.c.

DNM until tested.

pfl committed:
net: ip: Kconfig: Require mem pool to be configured

Require CONFIG_HEAP_MEM_POOL_SIZE to be set before selecting TCP2.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
samples: net: echo_client: Overlay file for TCP2

Set memory pool size and select the TCP2 stack in the overlay file.
Run cmake with argument '-DCONF_FILE="prj.conf overlay.tcp2"' to
use the overlay file.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -371,10 +371,12 @@ config NET_TCP1
+if HEAP_MEM_POOL_SIZE > 0
 config NET_TCP2
 	bool "Experimental TCP stack"
 	help
```
We could do this also by `depends on HEAP_MEM_POOL_SIZE > 0`

pfl commented:
Some more IPv4/IPv6 dependencies remain, will make an attempt on them during the next few days.

jukkar commented on samples/net/sockets/echo_client/prj.conf:
```diff
@@ -2,13 +2,16 @@
 CONFIG_NETWORKING=y
 CONFIG_NET_UDP=y
 CONFIG_NET_TCP=y
+#CONFIG_NET_TCP2=y
```
Please create an overlay config file `overlay-tcp2.conf` for these settings.

pfl replied:
Done.

jukkar reviewed (changes requested):
Looks good, couple comments only.

ozhuraki reviewed (changes requested):
@pfl

I suggest to split this pull request into 2:

-	Handling the heap. I think the proper way is to convert it into slab
-	Header access

For the header access, I would like the change to be postponed after we add IPv6 support.

pfl commented:
>     * Handling the heap. I think the proper way is to convert it into slab

This part should be done as in net_tcp_alloc() in the current tcp.c TCP stack.

tbursztyka reviewed (commented):
besides jukka's comments, lgtm.

pfl committed:
net: ip: Helper functions for IP payload length

Create helper functions for returning IP payload length. Sort
out the IP version from the net_pkt so that the caller has to
use only one function in a version agnostic way.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
subsys: net: tcp2: Use tcp.c checksumming function

Use the already existing checksumming function from tcp.c.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: tcp2: Use available function for finding TCP header

Use net_pkt_ip_hdr_len() and net_pkt_ip_opts_len() to determine the
starting offset of the TCP header. Ensure that the TCP header size is
continuous by setting net pkt access to the size of the TCP header in
net_pkt_get_data().

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: tcp2: Use existing helper functions

Use existing helper functions for fetching IP headers and their
lengths. Also take into account IPv4 header options and IPv6
extension header lengths.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: tcp2: Use existing functions to adjust IP length

Use existing functions that can adjust both IPv4 and IPv6 payload
lengths.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
net: tcp2: Remove unused defines

Remove unused defines since all users of these have bee updated.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>
